Sources/Purchasing/CachingTrialOrIntroPriceEligibilityChecker.swift

```diff
@@ -0,0 +1,107 @@
+//
+//  Copyright RevenueCat Inc. All Rights Reserved.
+//
+//  Licensed under the MIT License (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      https://opensource.org/licenses/MIT
+//
+//  CachingTrialOrIntroPriceEligibilityChecker.swift
+//
+//  Created by Nacho Soto on 10/27/22.
+
+import Foundation
+
+// swiftlint:disable type_name
+
+/// `TrialOrIntroPriceEligibilityCheckerType` decorator that adds caching behavior on each request.
+final class CachingTrialOrIntroPriceEligibilityChecker: TrialOrIntroPriceEligibilityCheckerType {
+
+    private let checker: TrialOrIntroPriceEligibilityCheckerType
+    private let notificationCenter: NotificationCenter
+
+    private let cache: Atomic<[String: IntroEligibility]> = .init([:])
+
+    private var observation: NSObjectProtocol!
```

Comment:
Hmm I'm probably missing something but should this be `private let observation: NSObjectProtocol`?

Reply:
The observation requires escaping `self` in the closure, and that's not fully initialized without setting this value. This is solved in Kotlin with `lateinit` but Swift doesn't have that.

Reply:
Ahh ok, thanks for explaining!

Reply:
this might not be obvious to a maintainer, and they might be inclined to try to spend some time getting rid of the IUO. Maybe we could add a comment explaining why this is needed?

Reply:
It's a normal pattern but yeah I can add it.

Reply:
Done.